### PR TITLE
Exercise03/Lara (resulting count of 1306)

### PR DESCRIPTION
**Pull request by laraasphh**

### CODE REVIEW

File: Exercises/LaraBachmann/exercises_03.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        return ', '.join(str(card) for card in self.cards)
+
+    def __len__(self):
+        return len(self.cards)
+            
+# PART 2:
+# Create a second class that represents a deck of cards usable for Skat -- it should only contain cards from 7 upwards.
+# It should offer all the same functionality of the first class.
+
+class Skat(FrenchDeck):
```
Comment (PixelWelt) [inline Exercises/LaraBachmann/exercises_03.py:33]: nice use for inheritance

File: Exercises/LaraBachmann/exercises_03.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+# I should be able to iterate over all cards in the deck.
+# Printing a cards string representation should give me a nice,
+# readable description of that card.
+
+class FrenchDeck:
+    suits = ['Diamonds', 'Hearts', 'Spades', 'Clubs']
+    ranks = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'Jack', 'Queen', 'King', 'Ass']
+
+    def __init__(self):
+        self.cards = [f'\n{rank} of {suit}' for suit in self.suits for rank in self.ranks]
```
Comment (PixelWelt) [inline Exercises/LaraBachmann/exercises_03.py:15]: I would've used a class for creating a Card but this is fine

File: Exercises/LaraBachmann/exercises_03.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    def __init__(self):
+        self.ranks = ['7', '8', '9', '10', 'Jack', 'Queen', 'King', 'Ass']
+        self.cards = [f'\n{rank} of {suit}' for suit in FrenchDeck.suits for rank in self.ranks]
+
+# Write some code to test the functionality of both kinds of decks. (You can use `assert` to make sure your classes behave the way you expect them to.)
+
+french_deck = FrenchDeck()
+print(f'Length of the French Deck: {len(french_deck)}')
+assert len(french_deck) == 52
```
Comment (PixelWelt) [inline Exercises/LaraBachmann/exercises_03.py:43]: take a look at pytest

File: Exercises/LaraBachmann/exercises_03.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    valid_numbers = []
+    valid_counter = 0
+
+    for number in range(lower_bound, upper_bound + 1):
+        number_str = str(number)
+        is_valid = True
+        has_pair = False
+        
+        for position in range(len(number_str) - 1):
+            if number_str[position] > number_str[position + 1]:
```
Comment (PixelWelt) [inline Exercises/LaraBachmann/exercises_03.py:78]: you could just apply sort to a copy of the string to just have one check

File: Exercises/LaraBachmann/exercises_03.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                is_valid = False
+                break
+
+        if is_valid:    
+            for position in range(len(number_str) - 1):
+                if number_str[position] == number_str[position + 1]:
+                    if  (position == 0 or number_str[position] != number_str[position - 1]) and \
+                        (position + 2 >= len(number_str) or number_str[position + 1] != number_str[position + 2]):
+                        has_pair = True
```
Comment (PixelWelt) [inline Exercises/LaraBachmann/exercises_03.py:87]: nice

**Review by PixelWelt — APPROVED** [review]
Nice Work

File: Exercises/LaraBachmann/exercises_03.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+# - 112233 is a valid number. At least one group of two is fulfilled, there is no maximum to the number of such groups.
+#
+# run your function with the lower bound `134564` and the upper bound `585159`. Note the resulting count
+# in your pull request, please.
+
+def validnumbers(lower_bound, upper_bound):
+    valid_numbers = []
+    valid_counter = 0
+
+    for number in range(lower_bound, upper_bound + 1):
```
Comment (LinusL2005) [inline Exercises/LaraBachmann/exercises_03.py:72]: Die upper_bound soll exclusive sein. Lass einfach der +1 weg. Sonst nen schöner Code
```suggestion
    for number in range(lower_bound, upper_bound):
```

**Pull request merged**